### ISSUE TITLE
Use predict to forecast timeseries Keras R

### ISSUE THREAD

**Issue by nakessien** [rstudio/keras#373]
Hi,

I'm an absolute newbie to deep learning, and however have managed to make the code (see below), which I saw online, work as in build the model and compile. I wish to know how to make it predict 24 steps ahead (i.e. 24 hours ahead), as this is the main task for me. Any help will be greatly appreciated and apologies if I haven't asked the question correctly.

`generator <- function(data, lookback, delay, min_index, max_index,`
                      `shuffle = FALSE, batch_size = 128, step = 6) {`
 `if (is.null(max_index))`
    `max_index <- nrow(data) - delay - 1`
 ` i <- min_index + lookback`
  `function()` `{
    if (shuffle) {
      rows <- sample(c((min_index+lookback):max_index), size = batch_size)
    } else {`
      `if (i + batch_size >= max_index)
        i <<- min_index + lookback
      rows <- c(i:min(i+batch_size-1, max_index))
      i <<- i + length(rows)
    }`
    
    samples <- array(0, dim = c(length(rows), 
                                lookback / step,
                                dim(data)[[-1]]))
    targets <- array(0, dim = c(length(rows)))
    
    for (j in 1:length(rows)) {
      indices <- seq(rows[[j]] - lookback, rows[[j]]-1, 
                     length.out = dim(samples)[[2]])
      samples[j,,] <- data[indices,]
      targets[[j]] <- data[rows[[j]] + delay,2]
    }            
    
    `list(samples, targets)
  `}
}`

`lookback <- 240
step <- 1
delay <- 24
batch_size <- 128`

`train_gen <- generator(
  data,
  lookback = lookback,
  delay = delay,
  min_index = 1,
  max_index = 5000,
  shuffle = TRUE,
  step = step, 
  batch_size = batch_size
)`

`val_gen = generator(
  data,
  lookback = lookback,
  delay = delay,
  min_index = 5001,
  max_index = 7000,
  step = step,
  batch_size = batch_size
)`

`test_gen <- generator(
  data,
  lookback = lookback,
  delay = delay,
  min_index = 7001,
  max_index = NULL,
  step = step,
  batch_size = batch_size
)`

`# How many steps to draw from val_gen in order to see the entire validation set
val_steps <- (7000 - 5001 - lookback) / batch_size`

`# How many steps to draw from test_gen in order to see the entire test set
test_steps <- (nrow(data) - 7001 - lookback) / batch_size`

`library(keras)
evaluate_naive_method <- function() {
  batch_maes <- c()
  for (step in 1:val_steps) {
    c(samples, targets) %<-% val_gen()
    preds <- samples[,dim(samples)[[2]],2]
    mae <- mean(abs(preds - targets))
    batch_maes <- c(batch_maes, mae)
  }`
  `print(mean(batch_maes))
}`

`evaluate_naive_method()`

`speed_mae<-0.21 * std[[3]]`

`#Basic ML`
`library(keras)`

`model <- keras_model_sequential() %>% 
  layer_gru(units=64, dropout=0.2, recurrent_dropout = 0.5, return_sequences=TRUE, input_shape = list(NULL, dim(data)[-1])) %>% 
  layer_gru(units = 128, dropout=0.1, recurrent_dropout = 0.5, activation = "relu") %>% 
  layer_dense(units = 1)`

`model %>% compile(
  optimizer = optimizer_rmsprop(),
  loss = "mae"
)`

`history <- model %>% fit_generator(
  train_gen,
  steps_per_epoch = 10,
  epochs = 40,
  validation_data = val_gen,
  validation_steps = val_steps
)`
`plot(history)`

`pred_ <- predict_classes(object = model, x = test_gen()) %>%  as.vector()`

**Comment by sanjmeh** [comment]
I am having trouble making the `generator` function work (this function is quoted in the the book: Deep learning in R by @jjallaire ). Hopefully Mr. Allaire will step in to clarify.

The function fails with an error.

```
Error in seq.default(rows[[j]] - lookback, rows[[j]], length.out = dim(samples)[[2]]) : 
  argument 'length.out' must be of length 1
```

Trying to understand how the function really works (since it is a little complicated), I tried step by step tracing variables inside it.

The part that is causing the function to fail is:
```
indices <- seq(rows[[j]] - lookback, rows[[j]]-1, 
                 length.out = dim(samples)[[2]])
```

I have attached a [short screen cas](https://youtu.be/Iof4N1-c4ZM)t of the trace run of the function created with following parameters:

`generator(dtx,lookback = 40,delay = 1,batch_size = 128,step=1,min_index = 1,max_index = 20000)
`

You can take any other parameters (I tried with a few more combinations) and they all fail at the same line that creates a list instead of a matrix perhaps due to indices not getting integer values.

**Comment by skeydan** [comment]
Hello,

can you post the exact code necessary to reproduce this please?

**Comment by sanjmeh** [comment]
Sure. Here it is.
```
generator <- function(data, lookback, delay, min_index, max_index,
                      shuffle = FALSE, batch_size = 128, step = 6) {
  if (is.null(max_index)) max_index <- nrow(data) - delay - 1
  i <- min_index + lookback
  function() {
    if (shuffle) {
      rows <- sample(c((min_index+lookback):max_index), size = batch_size)
    } else {
      if (i + batch_size >= max_index)
        i <<- min_index + lookback
      rows <- c(i:min(i+batch_size, max_index))
      i <<- i + length(rows)
}
    samples <- array(0, dim = c(length(rows),
                                lookback / step,
                                dim(data)[[-1]]))
    targets <- array(0, dim = c(length(rows)))
    for (j in 1:length(rows)) {
      indices <- seq(rows[[j]] - lookback, rows[[j]],
                     length.out = dim(samples)[[2]])
      samples[j,,] <- data[indices,]
      targets[[j]] <- data[rows[[j]] + delay,2]
    }
    list(samples, targets)
  }
}
```

And I invoke the above function using the following calls.

```
g2 <- generator(dtx,lookback = 40,delay = 1,batch_size = 128,step=1,min_index = 1,max_index = 20000)
g2()
```
Error received:
`Error in seq.default(rows[[j]] - lookback, rows[[j]], length.out = dim(samples)[[2]]) : 
  argument 'length.out' must be of length 1`

**Comment by skeydan** [comment]
The input to `generator` has to be a matrix (call `as.matrix` on the dataframe before)

**Comment by sanjmeh** [comment]
Ah ok ! Many thanks @skeydan

**Comment by skeydan** [comment]
You're welcome!
